File: app/src/main/java/com/ghrc/picpay/adapter/UserAdapter.java
package com.ghrc.picpay.adapter;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.Filter;
import android.widget.Filterable;
import android.widget.ImageView;
import android.widget.TextView;

import com.ghrc.picpay.R;
import com.ghrc.picpay.model.User;
import com.ghrc.picpay.util.ButtonHackClick;
import com.ghrc.picpay.util.CircleTransform;
import com.squareup.picasso.Picasso;

import java.util.ArrayList;


/**
 * Created by Guilherme on 27/08/2017.
 */

public class UserAdapter extends RecyclerView.Adapter<UserAdapter.MyViewHolder> implements Filterable {

    private Context context;
    private ArrayList<User> mListUser;
    private ArrayList<User> mFilteredList;
    private LayoutInflater mLayoutInflater;
    private ButtonHackClick button;

    public UserAdapter(Context context, ArrayList<User> mListUser, ButtonHackClick b) {
        this.context = context;
        this.mListUser = mListUser;
        this.mLayoutInflater = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        this.button = b;
    }

    @Override
    public MyViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return new MyViewHolder(mLayoutInflater.inflate(R.layout.item_user,parent,false));
    }

    public void addListItem(User u, int position)
    {
        mListUser.add(u);
        notifyItemInserted(position);
    }


    @Override
    public void onBindViewHolder(MyViewHolder holder, int position) {
        holder.tvName.setText(mListUser.get(position).getName() +  " | " + mListUser.get(position).getUsername());
        holder.tvIdUser.setText(mListUser.get(position).getId());
        Picasso.with(context).load(mListUser.get(position).getImg()).transform(new CircleTransform()).into(holder.ivPhoto);
    }

    @Override
    public int getItemCount() {
        return mListUser.size();
    }

    @Override
    public Filter getFilter() {
        return new Filter() {
            @Override
            protected FilterResults performFiltering(CharSequence charSequence) {
                String charString = charSequence.toString();
                if (charString.isEmpty()) {
                    mFilteredList = mListUser;
                } else {
                    ArrayList<User> filteredList = new ArrayList<>();
                    for (User user : mListUser) {
                        if (user.getUsername().contains(charSequence) || user.getName().contains(charSequence)) {
                            filteredList.add(user);
                        }
                    }
                    mFilteredList = filteredList;
                }
                FilterResults filterResults = new FilterResults();
                filterResults.values = mFilteredList;
                return filterResults;
            }

            @Override
            protected void publishResults(CharSequence charSequence, FilterResults filterResults) {
                mFilteredList = (ArrayList<User>) filterResults.values;
                notifyDataSetChanged();
            }
        };
    }

    class MyViewHolder extends RecyclerView.ViewHolder {
        ImageView ivPhoto;
        TextView tvName;
        TextView tvIdUser;
        Button btnEnviarPag;

        MyViewHolder(View itemView) {
            super(itemView);
            ivPhoto = (ImageView) itemView.findViewById(R.id.photo);
            tvName = (TextView) itemView.findViewById(R.id.name);
            tvIdUser = (TextView) itemView.findViewById(R.id.id_user);
            btnEnviarPag = (Button) itemView.findViewById(R.id.btnSendPag);
            btnEnviarPag.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    button.onClickListener(v, getLayoutPosition());
                }
            });
        }
    }

}

File: app/src/main/java/com/jvtnascimento/picpay/adapters/UserAdapter.kt
package com.jvtnascimento.picpay.adapters

import android.content.Context
import android.content.Intent
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.jvtnascimento.picpay.R
import com.jvtnascimento.picpay.models.User
import kotlinx.android.synthetic.main.item_user.view.*
import android.widget.Filter
import android.widget.Filterable
import com.jvtnascimento.picpay.dagger.modules.GlideApp
import com.jvtnascimento.picpay.ui.views.PaymentActivity


class UserAdapter(private val items : ArrayList<User>, private val context: Context) :
    RecyclerView.Adapter<UserAdapter.ViewHolder>(),
    Filterable {


    var onClick: (User) -> Unit = {}

    internal var filterListResult: ArrayList<User> = ArrayList()

    init {
        this.filterListResult = items
    }

    override fun getItemCount(): Int {
        return filterListResult.size
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        return ViewHolder(LayoutInflater.from(context).inflate(R.layout.item_user, parent, false))
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val user = filterListResult[position]

        GlideApp.with(context)
            .load(user.img)
            .placeholder(R.color.imageViewBackground)
            .into(holder.userPicture)

        holder.userName.text = user.name
        holder.userUsername.text = user.username

        holder.itemView.setOnClickListener {
            this.onClick(user)
        }
    }

    override fun getFilter(): Filter {
        return object: Filter() {
            override fun performFiltering(charString: CharSequence?): FilterResults {
                val charSearch: String = charString.toString()

                if (charSearch.isEmpty())
                    filterListResult = items
                else {
                    val resultList = ArrayList<User>()
                    for (row in items) {
                        if (row.name.toLowerCase().contains(charSearch.toLowerCase()))
                            resultList.add(row)
                    }

                    filterListResult =  resultList
                }

                val filterResults: FilterResults = Filter.FilterResults()
                filterResults.values = filterListResult
                return filterResults
            }

            override fun publishResults(charSequence: CharSequence?, filterResults: FilterResults?) {
                filterListResult = filterResults!!.values as ArrayList<User>
                notifyDataSetChanged()
            }
        }
    }

    class ViewHolder (view: View) : RecyclerView.ViewHolder(view) {
        val userName = view.userName!!
        val userUsername = view.userUsername!!
        val userPicture = view.userPicture!!
    }
}